lib/utils.js

```diff
@@ -11,31 +12,43 @@ function guid() {
   return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
 }
 
+/**
+ * Remove the specified module from the require cache.
+ * @param {string} moduleName 
+ */
 function purgeCache(moduleName) {
-  searchCache(moduleName, function (mod) {
+  return searchAndProcessCache(moduleName, function (mod) {
     delete require.cache[mod.id];
-  });
-  _.forEach(_.keys(module.constructor._pathCache), function(cacheKey) {
-    if (cacheKey.indexOf(moduleName)>0) {
-      delete module.constructor._pathCache[cacheKey];
-    }
+  })
+  .then(() => {
+    _.forEach(_.keys(module.constructor._pathCache), function(cacheKey) {
+      if (cacheKey.indexOf(moduleName)>0) {
+        delete module.constructor._pathCache[cacheKey];
+      }
+    });
+    return BbPromise.resolve();
   });
 }
 
-function searchCache(moduleName, callback) {
-  let mod = require.resolve(moduleName);
-  if (mod && ((mod = require.cache[mod]) !== undefined)) {
+function searchAndProcessCache(moduleName, processor) {
+  let mod_src = require.resolve(moduleName);
+  const visitedModules = [];
+  if (mod_src && ((mod_src = require.cache[mod_src]) !== undefined)) {
     (function traverse(mod) {
-      _.forEach(mod.children, function (child) {
-        traverse(child);
-      });
-      callback(mod);
-    }(mod));
+      if (!_.includes(visitedModules, mod)) {
+        visitedModules.push(mod);
+        _.forEach(mod.children, function (child) {
+          traverse(child);
```

Comment:
To prevent stackoverflows I'll replace the recursive implementation with a sequential one (breadth or depth first does not matter here).